### DISCUSSION TITLE
Differentiate eigenvalues

### DISCUSSION THREAD

**Discussion by andrea993** [autodiff/autodiff#368] [Q&A]
I need to differentiate eigenvalues evaluated with Eigen.
I'm trying  the following example

 ```
 
dual eigRealCalc(const dual &p)
{
    MatrixXdual M(2, 2);
    M << 2.0 * p, p * p, 4.0 * p, p;
    auto lambda =  M.eigenvalues();
    return lambda[0].real();
}

int main()
{
    auto lambdas_p = M.eigenvalues();
    std::cout << "Lambdas_p:\n"
              << lambdas_p << std::endl;*/

    dual x = 3.141;
    dual y = eigRealCalc(x);
    //auto g = derivative(eigRealCalc, wrt(x), at(x));

    std::cout << y << std::endl;
}

```
But I get the following error
```
/usr/include/eigen3/Eigen/src/Core/MathFunctions.h:989:39: error: call of overloaded ‘isfinite(const autodiff::detail::Dual<double, double>&)’ is ambiguous
[build]   989 |     return isfinite EIGEN_NOT_A_MACRO (x);
[build]       |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
[build] In file included from /usr/include/eigen3/Eigen/src/Core/util/Macros.h:679,
[build]                  from /usr/include/eigen3/Eigen/Core:19:
[build] /usr/include/c++/14.2.1/cmath:1134:3: note: candidate: ‘constexpr bool std::isfinite(long double)’
[build]  1134 |   isfinite(long double __x)
[build]       |   ^~~~~~~~
[build] /usr/include/c++/14.2.1/cmath:1130:3: note: candidate: ‘constexpr bool std::isfinite(double)’
[build]  1130 |   isfinite(double __x)
[build]       |   ^~~~~~~~
[build] /usr/include/c++/14.2.1/cmath:1126:3: note: candidate: ‘constexpr bool std::isfinite(float)’
[build]  1126 |   isfinite(float __x)
[build]       |   ^~~~~~~~
```

I read this issue https://github.com/autodiff/autodiff/issues/191 and if I understood the solution should be enable implicit casting to make isfinite() work properly 

I saw that now there is the macro `AUTODIFF_ENABLE_IMPLICIT_CONVERSION` that should do the work however also building with this macro the code still have the problem. 

How can I properly differentiate the eigenvalues function?

**Comment by NarakuK** [comment]
This new error is due to the Eigen library calling the isfinite function when calculating eigenvalues, but the dual type of the Autodiff library does not have a matching isfinite function overload, resulting in the compiler being unable to find a suitable isfinite function to handle Autodiff: Details: Parameters of type Dual<double, double>.
I personally think that overloading the isfinite function is a troublesome task. I may need to check the underlying code related to this in the Eigen library, or implement the feature value calculation function myself without using the functions in this library.

**Reply by andrea993** [reply]
Thanks for the reply, at the moment I'm using the old library and patch, but I think that a fix with the last eigen version will be useful for a lot of scientific applications

**Reply by NarakuK** [reply]
I know how to solve this problem now, and I am writing you a detailed explanation to clarify it. Please wait a moment.

**Comment by NarakuK** [comment]
**Step 1: Identify the root cause of the error**
The error originated from when Eigen called std: while calculating the eigenvalues internally isfinite ， But it cannot handle autodiff: Dual type, Eigen internally supports types such as double and float. We need to reload this isfinite function.

**Step 2: Define isfinite in the correct namespace**
Define the isfinite overload within the autodiff namespace to ensure that ADL (Parameter Dependency Lookup) is effective.
```
namespace autodiff::detail {
// ✅
bool isfinite(const Dual<double, double>& x) {
    return std::isfinite(x.val) && std::isfinite(x.grad);
}
}
```

**Step 3: Add type attribute support to Eigen**
By specializing in Eigen: NumTraits tells Eigen how to handle dual type numerical checks.Eigen relies on NumTraces to obtain the basic properties of a type (such as whether it is complex or initialized), and after specialization, ensures that Eigen correctly handles custom types.
```
#include <Eigen/Core>

namespace Eigen {
template<>
struct NumTraits<autodiff::detail::Dual<double, double>> : NumTraits<double> {
    typedef autodiff::detail::Dual<double, double> Real;
    typedef autodiff::detail::Dual<double, double> NonInteger;
    typedef autodiff::detail::Dual<double, double> Literal;
    enum {
        IsComplex = 0,
        IsInteger = 0,
        IsSigned = 1,
        RequireInitialization = 1,
        ReadCost = 1,
        AddCost = 3,
        MulCost = 3
    };
};
}
```
**Step 4: Modified complete code**
```
#include <iostream>
#include <autodiff/forward/dual.hpp>
#include <autodiff/forward/dual/eigen.hpp>
#include <Eigen/Eigenvalues>
#include <cmath>

using namespace autodiff;
using namespace Eigen;

namespace autodiff::detail {
// ✅
bool isfinite(const Dual<double, double>& x) {
    return std::isfinite(x.val) && std::isfinite(x.grad);
}
}

// ✅
template<>
struct Eigen::NumTraits<autodiff::detail::Dual<double, double>> : NumTraits<double>
{
    typedef autodiff::detail::Dual<double, double> Real;
    typedef autodiff::detail::Dual<double, double> NonInteger;
    typedef autodiff::detail::Dual<double, double> Literal;
    enum
    {
        IsComplex = 0,
        IsInteger = 0,
        IsSigned = 1,
        RequireInitialization = 1,
        ReadCost = 1,
        AddCost = 3,
        MulCost = 3
    };
};

dual eigRealCalc(const MatrixXdual& M) {
    auto lambda = M.eigenvalues();
    return lambda[0].real();
}

int main() {
    dual x = 3.141;
    MatrixXdual M(2, 2);
    M << 2.0 * x, x * x, 4.0 * x, x;

    const dual y = eigRealCalc(M);
    std::cout << "Real part of first eigenvalue: " << y << std::endl;

    return 0;
}
```
Mine can run normally now.
My English is not very good, if there are any issues, I can communicate further.

**Reply by syangav** [reply]
Hello! Thanks for your detailed answer. However, I was unable to compile the complete code you provided in step 4. Error messages: no matching function for call to 'fabs, scalbn, isnan, copysign, isinf'. 

What version of Eigen and g++ are you using? Or do you have any idea what the root cause is? 

More: I want to differentiate through the singular values of a parametric matrix (using Eigen::JacobiSVD). So I started by looking for something similar for matrix eigenvalues. Any help is welcome!
